x-pack/auditbeat/module/system/socket/socket.go

```diff
@@ -0,0 +1,421 @@
+
+	netlink *sock.NetlinkSession
+	// TODO: Replace with process data collected in processes metricset
+	ptable    *sock.ProcTable
+	listeners *sock.ListenerTable
+
+	bucket    datastore.Bucket
+	lastState time.Time
+}
+
+// Socket represents information about a socket.
+type Socket struct {
+	Family       linux.AddressFamily
+	State        linux.TCPState
+	LocalIP      net.IP
+	LocalPort    int
+	RemoteIP     net.IP
+	RemotePort   int
+	Inode        uint32
+	Direction    sock.Direction
+	UID          uint32
+	Username     string
+	ProcessPID   int
+	ProcessName  string
+	ProcessError error
+}
+
+// newSocket creates a new socket out of a netlink diag message.
+func newSocket(diag *linux.InetDiagMsg) *Socket {
+	return &Socket{
+		Family:     linux.AddressFamily(diag.Family),
+		State:      linux.TCPState(diag.State),
+		LocalIP:    diag.SrcIP(),
+		LocalPort:  diag.SrcPort(),
+		RemoteIP:   diag.DstIP(),
+		RemotePort: diag.DstPort(),
+		Inode:      diag.Inode,
+		UID:        diag.UID,
+		ProcessPID: -1,
+	}
+}
+
+// Hash creates a hash for Socket.
+func (s Socket) Hash() uint64 {
+	h := xxhash.New64()
+	h.WriteString(s.LocalIP.String())
+	h.WriteString(s.RemoteIP.String())
+	h.WriteString(strconv.Itoa(s.LocalPort))
+	h.WriteString(strconv.Itoa(s.RemotePort))
+	h.WriteString(strconv.FormatUint(uint64(s.Inode), 10))
+	return h.Sum64()
+}
+
+func (s Socket) toMapStr() common.MapStr {
+	evt := common.MapStr{
+		"network": common.MapStr{
+			"type": s.Family.String(),
+		},
+		"state": s.State.String(),
+		"user": common.MapStr{
+			"uid": s.UID,
+		},
+	}
+
+	switch s.Direction {
+	case sock.Outgoing:
+		evt.Put("network.direction", "outbound")
+		evt.Put("source", common.MapStr{
+			"ip":   s.LocalIP,
+			"port": s.LocalPort,
+		})
+		evt.Put("destination", common.MapStr{
+			"ip":   s.RemoteIP,
+			"port": s.RemotePort,
+		})
+	case sock.Incoming:
+		evt.Put("network.direction", "inbound")
+		evt.Put("source", common.MapStr{
+			"ip":   s.RemoteIP,
+			"port": s.RemotePort,
+		})
+		evt.Put("destination", common.MapStr{
+			"ip":   s.LocalIP,
+			"port": s.LocalPort,
+		})
+	case sock.Listening:
+		evt.Put("network.direction", "listening")
+		evt.Put("destination", common.MapStr{
+			"ip":   s.LocalIP,
+			"port": s.LocalPort,
+		})
+	}
+
+	if s.Username != "" {
+		evt.Put("user.name", s.Username)
+	}
+
+	if s.ProcessName != "" {
+		evt["process"] = common.MapStr{
+			"pid":  s.ProcessPID,
+			"name": s.ProcessName,
+		}
+	}
+
+	if s.ProcessError != nil {
+		evt.Put("process.error", s.ProcessError.Error())
+	}
+
+	return evt
+}
+
+// New constructs a new MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+
+	config := defaultConfig
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	bucket, err := datastore.OpenBucket(bucketName)
```

Comment:
I'm struggling to justify storing the socket state to disk. My thinking is that it would be fine to send the full to state upon restart. The downside is that you don't get a closed notification for sockets that went away the Beat was turned off. The upside is that we don't have to utilize the database/disk with updates on every fetch that happens on a 1s interval (configurable).

Reply:
It doesn't store the sockets themselves to disk - the state is resent on restart. The bucket is only used to store the `lastState` when the last state was sent - so we know when next to send it.

We could do without it, but then state would be sent at irregular intervals, i.e. state update happen at `00:00`, `01:00`, `01:12` (restart), `02:12`, etc.).

Just fyi, I thought of using a `cron` syntax for the `state.period` config parameter that does not require a local store - but I'm concerned about the "thundering herd" problem.

Reply:
That's fine with me. Ignore my comment. I didn't examine closely enough what was being persisted.

Reply:
The more I think about removing the bucket the more I like it. We would not be getting state in perfectly regular intervals, but how much does it matter? We are still guaranteed to have a state update at most every `state.period`.

I've also been thinking of reducing the default frequency of state updates (right now it's 12 hours) - then it matters even less.

A lot can happen in 12 hours on a busy system. Sending say every hour would reduce the timespan to be queried by over 90% while not increasing the data size by the same amount (on a busy system where event documents outweigh state, much less). Every hour does not seem like a lot considering Metricbeat sends its metrics every 10 seconds by default. It's a classic trade-off between data size between query time, and in most cases we've chosen query time.

What do you think?

Reply:
I've removed it for now, happy to revert if we think it's good to have it.